src/main/java/it/unitn/ds1/Replica.java

```diff
@@ -451,43 +830,196 @@ private void onCrashMsg(CrashMsg msg) {
+        }
+    }
+
+    private void onWriteMsgReceivedMsg(WriteMsgReceivedMsg msg) {
+        if (this.pendingUpdateRequests.contains(msg.updateRequestId)) {
+            // No WriteMsg received, coordinator crashed
+            this.recordCoordinatorCrash(
+                String.format("missed WriteMsg for write req %d from %s",
+                msg.updateRequestId.index,
+                msg.updateRequestId.client.path().name()
+            ));
+        }
+    }
+
+    private void onWriteOkReceivedMsg(WriteOkReceivedMsg msg) {
+        if (!this.writeOks.remove(msg.writeMsgId)) {
+            // No WriteOk received, coordinator crashed
+            this.recordCoordinatorCrash(
+                String.format("missed WriteOk for epoch %d index %d",
+                msg.writeMsgId.epoch,
+                msg.writeMsgId.index
+            ));
+        }
+    }
+
+    /**
+     * Checks how much time has elapsed since the last message received
+     * from the coordinator. If too much has elapsed, then a crash is detected.
+     */
+    private void onHeartbetReceivedMsg(HearbeatReceivedMsg msg) {
+        long now = new Date().getTime();
+        long elapsed = now - this.lastContact;
+        if (elapsed > Delays.RECEIVE_HEARTBEAT_TIMEOUT) {
+            // Too much time has passed since last hearing from the coordinator
+            // The coordinator in crashed
+            this.recordCoordinatorCrash(
+                String.format("missed HeartbeatMsg: %d elapsed",
+                elapsed
+            ));
+        }
+    }
+
+    private void onElectionAckReceivedMsg(ElectionAckReceivedMsg msg) {
+        var pair = new AbstractMap.SimpleEntry<ActorRef, Integer>(getSender(), msg.msg.index);
+        // If the pair is still in the set, the replica who should have sent the
+        // ACK is probably crashed
+        if (!this.pendingElectionAcks.contains(pair)) {
+            return;
+        }
 
+        this.crashedReplicas.add(getSender());
+        // The election message should be sent again
+        ActorRef nextNode = getNextNode();
+        if (nextNode == getSelf()) {
+            // There's no other active replica, so this should become the
+            // coordinator
+            getSelf().tell(new SynchronizationMsg(), getSelf());
+        } else {
+            this.tellWithDelay(nextNode, msg);
+        }
+    }
+
+    private void onCoordinatorAckReceivedMsg(CoordinatorAckReceivedMsg msg) {
+        var pair = new AbstractMap.SimpleEntry<ActorRef, Integer>(getSender(), msg.msg.index);
+        // If the pair is still in the set, the replica who should have sent the
+        // ACK is probably crashed
+        if (!this.pendingCoordinatorAcks.contains(pair)) {
+            return;
+        }
+
+        // The coordinator message should be sent again
+        ActorRef nextNode = getNextNode();
+        if (nextNode == getSelf()) {
+            // There's no other active replica, so this should become the
```

Comment:
This is impossible. There are at least (replicas.size() / 2) + 1 replicas alive at any time.

Reply:
Avevo pensato al caso in cui ci fosse una sola replica fin dall'inizio, ma effettivamente in quel caso non si arriverebbe mai l'election protocol perché non potrebbe crashare, quindi si può tranquillamente togliere quella parte